torchtune/utils/_device.py

```diff
@@ -5,9 +5,50 @@
 # LICENSE file in the root directory of this source tree.
 
 import os
-from typing import Optional
+from typing import Any, Optional, cast
 
 import torch
+if torch.xpu.is_available():
+    import intel_extension_for_pytorch
+
+class _DeviceHandle:
+    """
+    This is a simple abstraction for computing devices,
+    which enables custom backends that implement CUDA-like
+    semantics.
+    """
+
+    def __init__(self, device: torch.device, backend: Any = None):
+        if backend is None:
+            try:
+                self.__backend = getattr(torch, device.type)
+                self.__device = device
+            except AttributeError as exc:
+                raise AttributeError(
+                    f"Device '{device}' does not have a corresponding backend registered as 'torch.{device.type}'."
+                ) from exc
+        else:
+            self.__backend = backend
+
+    @classmethod
+    def from_device(cls, device: torch.device) -> "_DeviceHandle":
+        """
+        Return an device handle corresponding to the device, and through this handle,
+        operations with the same semantics as CUDA can be performed on the device.
+        Just return torch.cuda if the device is cuda to make attribute-access faster.
+        Custom backend must first register a module with the same name with {device.type} on torch.
+        """
+        if device.type == "cuda":
+            return cast(_DeviceHandle, torch.cuda)
+        return cls(device)
+
+    def __getattr__(self, __name: str) -> Any:
```

Comment:
a quick docstring here would be helpful, something along the lines of "retrieve attribute from correct backend based on registered device"

Reply:
Thank you for your suggestion! I will add docstring to it.